[receiver/vcenter] Adds vCenter VSAN Host metrics

BominRahmani opened the pull request:
**Description:** <Describe what has changed.>
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
The following PR adds VSAN host metric collection. The following metrics are collected 
```
vcenter.host.vsan.throughput          (direction={read/write})
vcenter.host.vsan.iops                (direction={read/write})
vcenter.host.vsan.congestions
vcenter.host.vsan.cache.hit_rate
vcenter.host.vsan.latency.avg         (direction={read/write})
```

**Link to tracking Issue:** #33556

**Testing:** <Describe what testing was performed and which tests were added.>
Added mock data fields in order to test vsan host metric scraping.  Updated golden test file output.

**Documentation:** <Describe the documentation added.>
Updated documentation using `make generate`

StefanKurek commented on receiver/vcenterreceiver/client.go:
```diff
@@ -314,3 +336,250 @@ func (vc *vcenterClient) PerfMetricsQuery(
+			if vsanErr != nil {
+				return nil, vsanErr
+			}
+			continue
```
This continue is probably fine (compared to returning "nil, &allResults"). My only concern here is that we don't have a test for this situation right now. To be fair, we also don't have a test for the happy path of multiple clusters either for vSAN as well. It would be a bit of a PITA, but it might be worth modifying the scrape tests/results to have a 2nd cluster with a single resource pool/host/VM.

Really....this would normally be handled in the client tests with the simulator, but the vSAN simulator isn't working.

BominRahmani replied:
I can modify the scrape tests, yea its a shame this won't be handled with the govmomi simulator, but I think this is a close second best alternative.

StefanKurek replied:
@BominRahmani Soooo, maybe you should take a look at the VM/Cluster PRs that I have out. I no longer am passing in a list of clusters, and this continue isn't actually needed...

StefanKurek replied:
resolve

StefanKurek commented on receiver/vcenterreceiver/client.go:
```diff
@@ -314,3 +336,250 @@ func (vc *vcenterClient) PerfMetricsQuery(
+}
+
+// vsanQueryByCluster performs a VSAN query for a specific cluster
+func (vc *vcenterClient) vsanQueryByCluster(
```
I can guess why you took the vSANDriver nil check out of this function, but this function should be able to stand on its own. So I would put it back here.

StefanKurek replied:
Again, in my other PRs this function doesn't even exist...

StefanKurek commented on receiver/vcenterreceiver/client.go:
```diff
@@ -314,3 +336,250 @@ func (vc *vcenterClient) PerfMetricsQuery(
+
+	rawResults, err := vc.vsanDriver.VsanPerfQueryPerf(ctx, clusterRef, querySpec)
+	if err != nil {
+		return nil, fmt.Errorf("problem retrieving %s VSAN metrics for cluster %s: %w", queryType, clusterRef.Value, err)
```
vSAN

StefanKurek commented on receiver/vcenterreceiver/client.go:
```diff
@@ -314,3 +336,250 @@ func (vc *vcenterClient) PerfMetricsQuery(
+	for _, rawResult := range rawResults {
+		metricResults, err := vc.convertVSANResultToMetricResults(rawResult)
+		if err != nil {
+			return &queryResults, fmt.Errorf("problem processing %s VSAN metrics for cluster %s: %w", queryType, clusterRef.Value, err)
```
vSAN

StefanKurek replied:
Also, this does simplify things from also checking for `metricResults == nil`, but you lose the ability to check for a specific problem UUID

BominRahmani replied:
Do you think the simplicity warrants the lack of extra information?

StefanKurek replied:
Extra information when a customer has a problem is often great to have.

StefanKurek commented on receiver/vcenterreceiver/client.go:
```diff
@@ -314,3 +336,250 @@ func (vc *vcenterClient) PerfMetricsQuery(
+// VSANQuery performs a VSAN query for the specified type across multiple clusters
+func (vc *vcenterClient) VSANQuery(
+	ctx context.Context,
+	queryType VSANQueryType,
```
Ok as I keep reviewing, I see that you will have individual functions calling this one. So you could make this one private if it's never intended to be called outside of the Client `vSANQuery`

StefanKurek replied:
resolve

StefanKurek commented on receiver/vcenterreceiver/scraper.go:
```diff
@@ -262,6 +266,22 @@ func (v *vcenterMetricScraper) scrapeHosts(ctx context.Context, dc *mo.Datacente
+
+	// Get all host vSAN metrics and store for later retrieval
+	clusterRefs := []*types.ManagedObjectReference{}
+	for _, compute := range v.scrapeData.computesByRef {
```
We could make a little helper function for this so it's not duplicated code.

StefanKurek replied:
In my new PRs this doesn't exist as we don't have to worry about passing in cluster anymore

StefanKurek replied:
resolve

StefanKurek commented on receiver/vcenterreceiver/scraper.go:
```diff
@@ -262,6 +266,22 @@ func (v *vcenterMetricScraper) scrapeHosts(ctx context.Context, dc *mo.Datacente
 		errs.AddPartial(1, fmt.Errorf("failed to retrieve perf metrics for HostSystems: %w", err))
 		return
```
I don't think you want to return here. Look at the VM function

StefanKurek commented on receiver/vcenterreceiver/metadata.yaml:
```diff
@@ -421,6 +443,51 @@ metrics:
+      if_enabled_not_set: "this metric will be enabled by default starting in release v0.106.0"
+  vcenter.host.vsan.latency.avg:
+    enabled: false
+    description: The host latency while accessing VSAN storage.
```
vSAN

StefanKurek commented on receiver/vcenterreceiver/metadata.yaml:
```diff
@@ -421,6 +443,51 @@ metrics:
+  vcenter.host.vsan.congestion:
+    enabled: false
+    description: The congestions of IOs generated by all vSAN clients in the host.
+    unit: "{congestions}"
```
because this isn't given as a counter i think we will want to convert this to `congestions/sec`

BominRahmani replied:
Can you explain this to me? I thought the congestions were a counter?

StefanKurek replied:
Let me rephrase. Because it isn't given as a sum counter (from the start of time) vs a delta counter (over a somewhat arbitrary interval).

StefanKurek replied:
resolve

StefanKurek commented on receiver/vcenterreceiver/metrics.go:
```diff
@@ -406,3 +406,57 @@ func (v *vcenterMetricScraper) recordVMPerformanceMetrics(entityMetric *performa
+			case "iopsWrite":
+				v.mb.RecordVcenterHostVsanOperationsDataPoint(pcommon.NewTimestampFromTime(*timestamp), value, metadata.AttributeVsanOperationTypeWrite)
+			case "throughputRead":
+				readRate := float64(value) / 20
```
we're now pulling the metric interval so you'll want to divide by that just like in the VMs

StefanKurek replied:
resolve

StefanKurek commented on receiver/vcenterreceiver/metrics.go:
```diff
@@ -406,3 +406,57 @@ func (v *vcenterMetricScraper) recordVMPerformanceMetrics(entityMetric *performa
+			case "latencyWrite":
+				v.mb.RecordVcenterHostVsanLatencyAvgDataPoint(pcommon.NewTimestampFromTime(*timestamp), value, metadata.AttributeVsanLatencyTypeWrite)
+			case "congestion":
+				v.mb.RecordVcenterHostVsanCongestionDataPoint(pcommon.NewTimestampFromTime(*timestamp), value)
```
Also divide this by interval I think

StefanKurek commented on receiver/vcenterreceiver/client.go:
```diff
@@ -314,3 +336,250 @@ func (vc *vcenterClient) PerfMetricsQuery(
+)
+
+// getLabelsForQueryType returns the appropriate labels for each query type
+func (vc *vcenterClient) getLabelsForQueryType(queryType VSANQueryType) []string {
```
Ok resolving based on what I saw later on.

BominRahmani commented on receiver/vcenterreceiver/client.go:
```diff
@@ -314,3 +336,250 @@ func (vc *vcenterClient) PerfMetricsQuery(
+) error {
+	faultErr := errors.Unwrap(err)
+	if faultErr == nil {
+		return fmt.Errorf("Problem retrieving %s vSan metrics: %w", queryType, err)
```
Good to know!

StefanKurek replied:
resolve

StefanKurek commented on receiver/vcenterreceiver/scraper.go:
```diff
@@ -264,17 +264,12 @@ func (v *vcenterMetricScraper) scrapeHosts(ctx context.Context, dc *mo.Datacente
 		errs.AddPartial(1, fmt.Errorf("failed to retrieve perf metrics for HostSystems: %w", err))
-		return
+	} else {
+		v.scrapeData.hostPerfMetricsByRef = results.resultsByRef
```
Because you added this line here, you want to remove the one below

StefanKurek replied:
@BominRahmani you still need to remove the line on 277

StefanKurek replied:
resolve

StefanKurek commented on receiver/vcenterreceiver/client.go:
```diff
@@ -464,17 +436,15 @@ func (vc *vcenterClient) vsanQueryByCluster(
-	clusterRefs []*vt.ManagedObjectReference,
 ) (*VSANQueryResults, error) {
-	return vc.vSANQuery(ctx, VSANQueryTypeHosts, clusterRefs)
+	return vc.vSANQuery(ctx, VSANQueryTypeHosts)
```
@BominRahmani unless you use the handleVSANError func inside here, you'll never turn any of the "lesser" errors to logs (same for the VM func)

StefanKurek replied:
resolve

StefanKurek commented on receiver/vcenterreceiver/metrics.go:
```diff
@@ -453,7 +453,8 @@ func (v *vcenterMetricScraper) recordHostVSANMetrics(vSANMetrics *VSANMetricResu
 			case "congestion":
-				v.mb.RecordVcenterHostVsanCongestionDataPoint(pcommon.NewTimestampFromTime(*timestamp), value)
+				congestionRate := float64(value) / float64(metric.Interval)
+				v.mb.RecordVcenterHostVsanCongestionDataPoint(pcommon.NewTimestampFromTime(*timestamp), int64(congestionRate))
```
Shouldn't be converting congestionRate to an int64. it should be a float in the metadata.yaml

StefanKurek replied:
resolve

StefanKurek commented on receiver/vcenterreceiver/testdata/metrics/expected-all-enabled.yaml:
```diff
@@ -2924,14 +2924,14 @@ resourceMetrics:
           - description: The congestions of IOs generated by all vSAN clients in the host.
             gauge:
               dataPoints:
-                - asInt: "1"
```
these are only 0 because you're converting to int. you could pick larger values in your mocked responses so they nicely divide by the interval (300) to get whole number results here, but not necessary

StefanKurek replied:
resolve

StefanKurek reviewed (changes requested):
Close. A few small changes left, but looking pretty good.

StefanKurek committed:
added vsan host metrics

remove golden.write

refactor client.go, introduce more abstraction, reduce duplication

Cleaned up client.go, use metric.interval for metrics collection

added changelog, removed clusterRefs

update int to float on congestion, update error handling calls
